python/taichi/lang/ast/ast_transformer.py

```diff
@@ -17,13 +17,34 @@
 from taichi.lang.util import is_taichi_class, to_taichi_type
 from taichi.types import (annotations, ndarray_type, primitive_types,
                           texture_type)
+from taichi.types.utils import is_integral
 
 if version_info < (3, 9):
     from astunparse import unparse
 else:
     from ast import unparse
 
 
+def boundary_type_cast_warning(expression):
+    expr_dtype = expression.ptr.get_ret_type()
+    if not is_integral(expr_dtype):
+        warnings.warn(
+            f"Casting range_for boundary values from {expr_dtype} to i32",
+            Warning)
+    elif expr_dtype == primitive_types.i64:
+        warnings.warn(
+            f"Casting range_for boundary values from {expr_dtype} to i32, which may cause numerical issues",
+            Warning)
+    elif expr_dtype == primitive_types.u64:
+        warnings.warn(
+            f"Casting range_for boundary values from {expr_dtype} to i32, which may cause numerical issues",
+            Warning)
+    elif expr_dtype == primitive_types.u32:
+        warnings.warn(
+            f"Casting range_for boundary values from {expr_dtype} to i32, which may cause numerical issues",
+            Warning)
```

Comment:
```suggestion
    if not is_integral(expr_dtype) or expr_dtype in [primitive_types.i64, primitive_types.u64, primitive_types.u32]:
        warnings.warn(
            f"Casting range_for boundary values from {expr_dtype} to i32, which may cause numerical issues",
            Warning)
```
What about making this simpler :-)